scipy/signal/signaltools.py

```diff
@@ -777,6 +777,88 @@ def convolve(in1, in2, mode='full', method='auto'):
     return correlate(volume, _reverse_and_conj(kernel), mode, 'direct')
 
 
+def circular_convolve(in1, in2, period):
+    """
+    Circular convolution of two 1-dimensional arrays.
+
+    Circular convolve `in1` and `in2` with given `period`.
+
+    Parameters
+    ----------
+    in1 : array_like, 1-D
+        First input.
+    in2 : array_like, 1-D
+        Second input. Should have the same number of dimensions as `in1`.
+    period : int
+        Period of the circular convolution.
+
+    Returns
+    -------
+    result : array, 1-D
+        A 1-dimensional array containing the result of the circular
+        convolution of `in1` with `in2`.
+
+    See Also
+    --------
+    convolve
+
+    Notes
+    -----
+    The (modulo-M) circular/cyclic/periodic convolution of period :math:`M`
+    of the two signals :math:`x[k]` and :math:`h[k]` is defined as
+
+    .. math::
+
+        y[k] = \sum_{\kappa=0}^{M-1} \tilde{x}[k - \kappa] \; \tilde{h}[\kappa]
+
+    where the periodic summations :math:`\tilde{x}[k]` and `\tilde{h}[\kappa]`
+    of :math:`x[k]` and :math:`x[k]` are defined as
+
+    .. math::
+
+        \tilde{x}[k] &= \sum_{m = -\infty}^{\infty} x[m \cdot M + k] \\
+        \tilde{h}[k] &= \sum_{m = -\infty}^{\infty} h[m \cdot M + k]
+
+    Examples
+    --------
+    Equivalence of circular and linear convolution:
+
+    >>> from scipy import signal
+    >>> a = np.ones(5)
+    >>> b = np.ones(5)
+    >>> circular_convolve(a, b, 5)
+    array([ 5.,  5.,  5.,  5.,  5.])
+    >>> np.convolve(a, b, mode='full')
+    array([ 1.,  2.,  3.,  4.,  5.,  4.,  3.,  2.,  1.])
+    >>> circular_convolve(a, b, 9)
+    array([ 1.,  2.,  3.,  4.,  5.,  4.,  3.,  2.,  1.])
+    """
+    in1 = _periodic_summation(in1, period)
+    in2 = _periodic_summation(in2, period)
+
+    return np.array([np.dot(np.roll(in1[::-1], k+1), in2)
```

Comment:
Shouldn't this be multiplying ffts instead of multiplication in a loop?

Reply:
The result of above calculation is indeed equivalent to multiplying the FFTs of both signals followed by an inverse FFT. However, both techniques will differ with respect to round-off errors, for an example see [this Jupyter notebook](http://nbviewer.jupyter.org/github/spatialaudio/digital-signal-processing-lecture/blob/master/nonrecursive_filters/quantization_effects.ipynb#Example---Comparison-of-round-off-noise-of-conventional-and-fast-convolution) and computational complexity. The aspect of complexity was already discussed above  

> Yes the numerical complexity will be high for long signals. However I wanted to stay in line with > the usual convention used for the familiy of convolve functions (e.g. numpy.convolve, conv in 
> MATLAB, ...) which implement the convolution by summation.

But there is also an additional aspect, the current implementation works seamless with different data types (e.g. integer, float, complex).

Reply:
Did you see all the changes in https://github.com/scipy/scipy/pull/5608 to speed up linear convolution?

Reply:
@endolith Yes I am aware of the achievable speedups by fast convolution, see for instance [my notebooks](http://nbviewer.jupyter.org/github/spatialaudio/digital-signal-processing-lecture/blob/master/nonrecursive_filters/fast_convolution.ipynb#Example---Numerical-complexity) on the topic. However, following the argumentation above I would be in favour of at least having the choice of using a straightforward or FFT-based implementation. I could replicate the functionality of the convole function?

Reply:
@spors Yeah, I would recommend using the same `method={'direct', 'fft', 'auto'}` functionality, and combining code between the two when possible.